design/runtime-parameters.md

```diff
@@ -0,0 +1,226 @@
+# Runtime parameters
+
+We need a way to take run-time inputs, giving pipeline authors the right amount of control / expressivity.
+YAML pipelines already accept parameters when used as part of a template.
+Runtime parameters are a natural evolution of that syntax.
+
+Scenarios:
+- Pipeline users can supply different values to tools and tasks at run time
+- Pipeline authors can control the types, ranges allowed, and defaults for run-time parameters
+- Pipeline authors can allow flexible, run-time control over stages/jobs to run, including adding/removing matrix legs
+- The pipeline itself can be altered at run time: pools, agentspecs, service connections, environments, and more
+- The runtime values can be used in template expressions (`${{ if }}`, for example), for instance so that jobs/steps can be dynamically selected
+
+## Syntax
+
+```yaml
+parameters:
+# this is a sequence, so the block may be repeated as many times as needed
+- name: string          # name of the parameter; required
+  displayName: string   # UI string
+  displayHint: enum     # see below
+  type: enum            # data type; see below
+  default: any          # default value; if no default, then the parameter MUST be given by the user at runtime
+  values: [ string ]    # allowed list of values (for some data types)
+```
+
+See [below](#back-compat-with-current-syntax) for details about how this coexists with existing `parameters` syntax.
+
+## User interactions
+
+### Run panel ("queue dialog")
+
+_TODO_
+
+### Comment triggers
+
+_TODO_
+
+### Editor experiences
+
+We can teach the editors (web, VS Code) how templates fit together.
+When a user's pipeline consumes a template, we can offer IntelliSense for the names and allowed values of parameters.
+*This is out of scope for v1.*
+
+## Data types
+
+Available data types:
+
+| Data type | Notes |
+|-----------|-------|
+| `string` | default data type if none is specified. if `values:` is specified, they become suggestions but **not** required (like a combobox)
+| `number` | may be restricted to `values:`, otherwise any number-like string is accepted
+| `boolean`
+| `object` | YAML serialization expected
+| `enum` | must be restricted to `values:` (like a selectbox)
+| `filePath`
+| `secureFile`
+| `pool`
+| `serviceConnection`
+| `environment`
+
+It's also useful to tighten up the editor experience and error messages for template parameters.
+Customers are already using templates to pass along step and job lists.
+These aren't very useful for run-time parameters: it's unclear how a user would be expected to set them.
+They're rendered like a plain `object` in the UI.
```

Comment:
What is the change you're making?

Reply:
More confusion between runtime and template parameters. Fixing.